graph-builder/src/registry.rs

```diff
@@ -37,24 +46,163 @@ impl Into<cincinnati::Release> for Release {
+                    .and_then(move |is_auth| {
+                        if !is_auth {
+                            Err("login failed".into())
+                        } else {
+                            println!("logged in!");
+                            Ok(dclient.set_token(Some(token.token())))
+                        }
+                    })
+            })
+        })
+}
+
 /// Fetches a vector of all release metadata from the given repository, hosted on the given
 /// registry.
-pub fn fetch_releases(registry: &str, repo: &str) -> Result<Vec<Release>, Error> {
-    let mut metadata = Vec::new();
-    for tag in fetch_tags(registry, repo)? {
-        metadata.push(Release {
-            source: format!(
-                "{}/{}:{}",
-                registry
-                    .trim_left_matches("https://")
-                    .trim_left_matches("http://"),
-                repo,
-                tag
-            ),
-            metadata: fetch_metadata(registry, repo, &tag)?,
+pub fn fetch_releases(
+    registry: &str,
+    repo: &str,
+    credentials_path: &Option<PathBuf>,
+) -> Result<Vec<Release>, Error> {
+    let mut tcore = Core::new()?;
+
+    let registry_host = trim_protocol(registry);
+
+    let mut client = dkregistry::v2::Client::configure(&tcore.handle())
+        .registry(registry_host)
+        .insecure_registry(false)
+        .read_credentials(get_credentials_or_empty_reader(credentials_path))
+        .build()
+        .map_err(|e| format_err!("{}", e))?;
+
+    let authenticated_client = tcore
+        .run(authenticate_client(
+            &mut client,
+            format!("repository:{}:pull", &repo),
+        ))
+        .map_err(|e| format_err!("{}", e))?;
+
+    let releases = authenticated_client
+        .get_tags(&repo, None)
+        .and_then(|tag| {
+            trace!("processing: {}:{}", &repo, &tag);
+
+            let tag_clone0 = tag.clone(); // TODO(steveeJ): is there a way to avoid this?
+            let manifest_kind_future = authenticated_client
+                .has_manifest(&repo, &tag, None)
+                .and_then(move |manifest_kind| match manifest_kind {
+                    Some(manifest_kind) => Ok(manifest_kind),
+                    None => {
+                        Err(format!("{}:{} doesn't have a manifest", &repo, &tag_clone0).into())
+                    }
+                })
+                .inspect(|manifest_kind| {
+                    trace!("manifest_kind: {:?}", manifest_kind);
+                });
+
+            let manifest_future = authenticated_client.get_manifest(&repo, &tag);
+
+            let layer_digests_future =
+                manifest_kind_future
+                    .join(manifest_future)
+                    .map(|(manifest_kind, manifest)| match manifest_kind {
+                        dkregistry::mediatypes::MediaTypes::ManifestV2S1Signed => {
+                            let m: dkregistry::v2::manifest::ManifestSchema1Signed =
```

Comment:
Is this intermediate variable necessary?

Reply:
It seems to be. When I rewrite one of the match arms as 

```rust
                        dkregistry::mediatypes::MediaTypes::ManifestV2S1Signed => Ok((
                            serde_json::from_slice(manifest.as_slice())?.get_layers(),
                            tag,
                        )),
```
the compiler prints:
```console
error[E0282]: type annotations needed                                                                                                        
   --> graph-builder/src/registry.rs:151:29                                                                                                  
    |                                                                                                                                        
151 |                             serde_json::from_slice(manifest.as_slice())?.get_layers(),                                                 
    |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type                                             
    |                                                                                                                                        
    = note: type must be known at this point                                                                                                 
                                                                                                                                             
error: aborting due to previous error                                                                                                        
                                                                                                                                             
For more information about this error, try `rustc --explain E0282`. 
```

Reply:
I was able to get rid of the assignments.